File: SCL/Math/差分/差分数组.py
if __name__ == "__main__":
    m,n = map(int,input().split())
    
    a = list(map(int,input().split()))
    
    b = [0] * (m + 10)
    
    def insert(l,r,c):
        b[l] += c
        b[r + 1] -= c
    
    for i in range(m):
        insert(i+1,i+1,a[i])
    
    
    for i in range(n):
        l,r,c = map(int,input().split())
        insert(l,r,c)
    
    for i in range(1,m + 1):
        b[i] += b[i - 1]
    for i in range(1,m+1):
        print(b[i],end = ' ')

File: SCL/Math/差分/差分数组 .cpp
#include <iostream>

using namespace std;

const int N = 1e5 + 10;
int a[N],b[N];

//第一次构造差分数组的时候，
//将给的数组看作是空的，这样就等价于在区间[i,i]上插入c了
//代码复用度会高
void insert(int l,int r,int c){
    b[l] += c;
    b[r + 1] -= c;
}


int main(){
    int m,n;
    scanf("%d%d",&m,&n);
    for(int i = 1;i <= m;i++) scanf("%d",&a[i]);
    
    for(int i = 1;i <= m;i++) insert(i,i,a[i]);

    for(int i = 0;i < n;i++){
        int l,r,c;
        scanf("%d%d%d",&l,&r,&c);
        insert(l,r,c);
    }   
    for(int i = 1; i <= m;i++) b[i] += b[i - 1];

    for(int i = 1;i <= m;i++) printf("%d ",b[i]);

    return 0;    
}